### ISSUE TITLE
Custom ModelValidator is not working for some types

### ISSUE THREAD

**Issue by evorine** [aspnet/Mvc#4490]
I'm trying to implement a xml based custom model validation.
But `IModelValidatorProvider.GetValidators` method won't invoke for some reasons for integer type properties.

*BTW: I'm using rc1-final version*

Here is what I did:
**`DemoModel`:**
```
public class DemoModel
{
    public string NotRequiredString { get; set; }

    public string RequiredString { get; set; }

    public int NotRequiredInteger { get; set; }

    public int RequiredInteger { get; set; }

    public int? NotRequiredNullableInteger { get; set; }

    public int? RequiredNullableInteger { get; set; }
}
```

**`IModelValidatorProvider`:**
```
public class DynamicModelValidatorProvider : IModelValidatorProvider
{
    public void GetValidators(ModelValidatorProviderContext context)
    {
        context.Validators.Add(new DynamicModelValidator());
    }
}
```

**`IModelValidator`:**
```
public class DynamicModelValidator : IModelValidator
{
    public bool IsRequired { get { return true; } }

    public IEnumerable<ModelValidationResult> Validate(ModelValidationContext context)
    {
        // context.Metadata.PropertyName
        // This method is not being called for integer typed properties
        // For example: RequiredInteger and NotRequiredInteger
    }
```

#### My request is something like:
`RequiredString=&NotRequiredString=&RequiredInteger=&NotRequiredInteger=&RequiredNullableInteger=&NotRequiredNullableInteger=`

#### Actual
`IModelValidatorProvider.GetValidators` won't invoke for every property of `DemoModel`. It ignores `int`'s (`RequiredInteger`, `NotRequiredInteger`).

#### Expected behavior
`IModelValidatorProvider.GetValidators` should be called for every property to validate them?

**Comment by evorine** [comment]
I changed my request data and set one of the ignored int property to some value (for example: 0).
And now my model validator is getting called for that property too.

Like:
`RequiredString=&NotRequiredString=&RequiredInteger=&NotRequiredInteger=0&RequiredNullableInteger=&NotRequiredNullableInteger=`

**Comment by dougbu** [comment]
You don't need a custom validator if you wish to require a non-`Nullable` value type. Use the `[BindRequired]` attribute on the relevant properties. That attribute actually does its work during model binding, not validation.

**Comment by Eilon** [comment]
@dougbu any idea why the validator wouldn't run, though?

**Comment by dougbu** [comment]
> any idea why the validator wouldn't run, though?

A guess: The validator won't be invoked if the property is already invalid. @evorine does the `ModelStateDictionary` contain an entry for `RequiredInteger` for example?

**Comment by evorine** [comment]
@dougbu 
You were right. It has already an error for `RequiredInteger`. It says **The value '' is invalid.**. I think it comes from `DefaultModelValidator`.

Let me explain what I'm trying to do here. I have a `DynamicValidationAttribute` attribute. I want to use a custom `IModelValidator` implementation if a model class has that attribute.
I need this because some of my model classes's data annotation rules will be reconfigurable. I mean, max lengths, requirements etc. rules will come from a xml file.

I tried to ignore `DefaultModelValidator` for classes marked with `DynamicValidationAttribute`'s but I could not find any way to do it.
For example I had this code before:
```
public class DynamicModelValidatorProvider : IModelValidatorProvider
{
    public void GetValidators(ModelValidatorProviderContext context)
    {
        if (context?.ModelMetadata?.ContainerType != null)
        {
            var dynamicAttributes = context.ModelMetadata.ContainerType.GetCustomAttributes(typeof(ßDynamicValidationAttribute), true);
            if (dynamicAttributes.Any())
            {
                context.Validators.Clear(); // Here I also hoped to remove other default validators. But it was already empty when this code hit :(
                context.Validators.Add(new DynamicModelValidator());
            }
        }
    }
}
```

But currently I removed the code above and switched my custom validator from `IModelValidator` to `IObjectModelValidator`.

As there is no documentation at the moment, I am unable to find the proper way. I think I need some advise here :sos: 

TL;DR: Is there any way to decide which validators will be used for classes based on their metadata while model validationing?

**Comment by dougbu** [comment]
> any way to decide which validators will be used for classes based on their metadata while model validationing?

I suggest instead that your custom validator itself (not a provider) should behave dynamically.

Also suggest your validator might work more as you'd like if it were added just for properties or parameters of type `DemoModel`. That way property-level errors inside `DemoModel` wouldn't prevent calls to your validator.

@rynowak any other thoughts?

**Comment by rynowak** [comment]
> Also suggest your validator might work more as you'd like if it were added just for properties or parameters of type DemoModel. That way property-level errors inside DemoModel wouldn't prevent calls to your validator.

The validation system won't call validators on `DemoModel` if there are property level errors - this is part of the validation system as a whole.

**Comment by evorine** [comment]
@dougbu @rynowak 

Thank you @dougbu. I did not notice it was a binding related thing before you mentioned it. I was trying to find the answers in model validation methods...

So if I get it right, my problem is that the binding system won't be silent about value typed properties who don't have any value int the request data. And this prevents invoking validation methods for those properties.

As I don't want to convert all of my value types into `Nullable<T>` class, I add my own `IModelBinder` implementation and make sure it's only called when my `DynamicValidationAttribute` is added for the container class. And this model binder will set `default()` for value typed parameters if they are empty. This way I won't have any entry in `ModelStateDictionary` because of empty values for value typed properties. And I think this way my validation methods will be called and they will be all be responsible for the validation system.
I hope I'm right here :pray: 

Basically something like this will fix my problem:
```
public class CustomModelBinder : IModelBinder
{
    public Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindingContext)
    {
        // Here I will add codes to check if the container class have required attribute...
        if (check attribute here)
        {
            if (bindingContext.ModelType == typeof(int))
            {
                var valueResult = bindingContext.ValueProvider.GetValue(bindingContext.ModelName);
                if (valueResult.FirstValue == "") return Task.FromResult(ModelBindingResult.Success(bindingContext.ModelName, default(int)));
            }
        }
        return ModelBindingResult.NoResultAsync;
    }
}
```

Is this a recommended way?

**Comment by evorine** [comment]
BTW I tested it right now and it worked for me. I'm closing this. But still value your opinions and want to know if this is not a very very very bad and dirty hack :)

**Comment by rynowak** [comment]
Yeah what you did here with the custom model binder seems fine. Note that this code will change somewhat significantly in RC2

**Comment by dougbu** [comment]
:+1: One small note: Add your `CustomModelBinder` before `SimpleTypeModelBinder`.
